app/controllers/observer_controller/indexes.rb

```diff
@@ -218,11 +218,11 @@ def map_observations
 
     unless locations.empty?
       # Eager-load corresponding locations.
-      @locations = Location.connection.select_rows(%(
-        SELECT id, name, north, south, east, west FROM locations
-        WHERE id IN (#{locations.keys.sort.map(&:to_s).join(",")})
-      )).map do |id, *the_rest|
-        locations[id.to_i] = MinimalMapLocation.new(id, *the_rest)
+      @locations = Location.
+                   where(id: locations.keys.sort).
+                   pluck(:id, :name, :north, :south, :east, :west).
+                   map do |id, *the_rest|
+        locations[id] = MinimalMapLocation.new(id, *the_rest)
       end
```

Comment:
Naturally I don't have the capacity to benchmark this, but from what Nimmo and I were seeing not too long ago in a not-altogether-different case, it is actually a good deal faster these days to instantiate all of the locations, and then grab the columns you need from the instances.  Admittedly, though, this could be *huge*.  I mean, for example, it is entirely possible that folks will try to map every single location in our database(!)  So it might be a good idea to benchmark this one properly.  And pay attention to memory use, as well as system/user time, this time around.

For clarity, I'm recommending comparing what you've written with:

`Location.where(id: locations.keys).map do |loc|
    locations[id] = MinimalMapLocation.new(id, loc.name, loc.north, loc.south, loc.east, loc.west)
end`

(Note that there is no reason to sort locations.keys... I don't think.  I guess it depends what mysql does on the database end.  Maybe if dbserver time is more precious than webserver time it really is better to sort it however mysql wants it sorted on the webserver end.  But how would we know what will make mysql's life easiest?  In the absence of any information, it stands to reason we should do nothing -- that is, leave locations.keys unsorted until we have some benchmark or documentation proving that it is better to sort it somehow.)

Reply:
How about limiting the number of locations this method will map to a sane value?

Reply:
(I'm waiting on a tow truck, wont make it to retreat tonight!)